### DISCUSSION TITLE
red coloured digit not recognized

### DISCUSSION THREAD

**Discussion by amzrhh** [jomjol/AI-on-the-edge-device#833] [Q&A]
Hi experts,
i am running an installation of AI on the edge to monitor the output of a rainwater store. The watermeter i am using has got only a digital counter. The problem i am struggling with is the first digit after the decimal dot, which is red coloured. (see attached picture.) There is never a match. I am thinking about to take the images in black and white but i haven't found an advice so i don't know now to do so. 
Many thanks in advance for a hint.
![Digital counter](https://user-images.githubusercontent.com/108123803/175507271-85f4ac58-43fa-4a66-987c-ae7f5063ec99.JPG)

**Comment by jomjol** [comment]
Taking the image in black and white is not possilbe as the the neural network expects a RGB image. The only way to improve is to add you type of image to the training.

**Comment by danieldownload** [comment]
Same here, Andrae/TECHEM Meter (Germany), only one row with 8 digits, the last 3 are in red color. Some got recognized, some not.

Any chance to improve it?

<img width="668" alt="Bildschirmfoto 2022-08-13 um 15 41 24" src="https://user-images.githubusercontent.com/26095164/184496846-c6fbbbc0-ea33-4926-96cf-00c579f86c21.png">

**Comment by jomjol** [comment]
I need trainings data - that will help

**Reply by danieldownload** [reply]
Ok, how many and in which format/type?

**Comment by jomjol** [comment]
https://github.com/jomjol/AI-on-the-edge-device/wiki/ROI-Configuration#1-get-images-into-the-next-training-round

**Reply by danieldownload** [reply]
I need to correct my POI's first, how can I run the setup process again without to remove and rewrite the sd card?

**Reply by ghost** [reply]
You do not need setup again. Go to "Configuration" and then to "Region of Interest (ROI)"->"Digital ROIs"

**Reply by danieldownload** [reply]
Super, I did a ROI correction and now lets see if it gets better. I noticed that the picture on Overview page is always overexposed/too bright while it's shown correctly on Reference Image site. After referencing the image again, it's shown correctly on Overview site. Same procedure is needed after next reboot, what I'm doing wrong here?

**Comment by amzrhh** [comment]
Hi experts,

there could  be another issue with the watermeter below, at least I 
struggled with it:
I assume the last digit is not a real "digital" one. It turns continouly 
slowly (from 0 to 1  for example) while passing 1000 Liter, in my case 
it could last about half a week.

Only when the digit it fully visible, the programm gets the right value. 
For me only 2 times a week I got the right values for some hours.

The red color problem I solved by exchanging the red digit with a black 
leading zero digit (exciting puzzle to reassemble the counter after 
disassembling), but I failed to find a solution for the quasi-analog 
operating last digit.
So I exchanged the watermeter with one of that type the brillant AI on 
the edge is trained on.


best regards,

August


Am 13.08.2022 um 15:45 schrieb danieldownload:

**Reply by danieldownload** [reply]
Hello August,

interesting point, in fact the last number is moving slowly but my understanding was that this is the normal behavior for such a meter. Never saw a meter changing like a flip-clock.

<img width="880" alt="Bildschirmfoto 2022-08-13 um 17 22 07" src="https://user-images.githubusercontent.com/26095164/184500556-3bbaed07-0d8d-4c3f-978a-a884e4bec4e8.png">

**Comment by ghost** [comment]
You could try the new 11.0.0 Version with the new models. I think it's more robust and you will have intermediate states. But you will need to reconfigure the digital ROIs in height.

**Reply by danieldownload** [reply]
With pleasure, how can I get V11?

**Reply by jomjol** [reply]
Published about 3 hours ago in the github repo. You need to choose also an appropriate tlite model (dig-class100..., or dig-cont...) for using the new feature in the settings.

**Reply by danieldownload** [reply]
Can I update via OTA?

**Reply by ghost** [reply]
Yes. Don't forget the html.zip to upload too.

**Reply by danieldownload** [reply]
Update done, overexposure on Overview page is still there, that's confusing. I run the model dig-cont_0560_s3_q.

<img width="789" alt="Bildschirmfoto 2022-08-13 um 19 01 50" src="https://user-images.githubusercontent.com/26095164/184503829-fa502804-59cd-4bbe-8fa2-cde71d326368.png">

<img width="778" alt="Bildschirmfoto 2022-08-13 um 19 05 28" src="https://user-images.githubusercontent.com/26095164/184503865-e5de2d0b-f643-47a0-8f9a-558792a469f1.png">

<img width="784" alt="Bildschirmfoto 2022-08-13 um 19 07 15" src="https://user-images.githubusercontent.com/26095164/184503918-126d34ab-1a65-4341-9626-9e63e2de951a.png">

**Reply by ghost** [reply]
Which values do you set for LEDIntension and Brightness? (Expert mode in configuration)
Here my values:
<img width="323" alt="image" src="https://user-images.githubusercontent.com/412645/184504567-84c87902-65f7-47e8-b087-13e0acbf8bda.png">

**Reply by danieldownload** [reply]
-1, like on Reference Image page.

<img width="567" alt="Bildschirmfoto 2022-08-13 um 19 31 16" src="https://user-images.githubusercontent.com/26095164/184504658-e2d7ab4c-bcff-4d24-b07a-aae2b3bc4ab2.png">

<img width="852" alt="Bildschirmfoto 2022-08-13 um 19 32 10" src="https://user-images.githubusercontent.com/26095164/184504689-57033430-8683-4c48-9914-dfaac241b6fb.png">

**Reply by ghost** [reply]
Your ROIs for the new models are a bit to high. I've used a setup like described in https://github.com/haverland/collectmeterdigits

**Reply by ghost** [reply]
Reduce the LEDIntensity. 5-10 is a good value if the complete picture is not to dark

**Reply by danieldownload** [reply]
The brightness is ok in Reference page, but on Overview page it's different (same like the digit squares), after a while the Overview page shows also the correct brightness in the image.

So the brightness itself seems not to be the prob. Right now the Overview page shows the correct image (brightness).

Gonna set the ROIs now for the new model.

**Reply by danieldownload** [reply]
Btw. Fine Alignment is never saved.

<img width="780" alt="Bildschirmfoto 2022-08-13 um 19 40 44" src="https://user-images.githubusercontent.com/26095164/184504944-3e77fb79-786a-4131-9659-4670446ed9c4.png">

**Reply by ghost** [reply]
Yes. It will only be used for adjusting the reference image. Later the alignement process will be used for correction.

**Reply by danieldownload** [reply]
Checked the alignment, it came over from old FW and it looks ok for me. Also the Overview page shows everything correct except the half number on the last position.

The position of the number is fluctuating, so it's never exactly on the same position after a turn. That's why the 6 is a bit out of the box.


<img width="1145" alt="Bildschirmfoto 2022-08-13 um 19 54 06" src="https://user-images.githubusercontent.com/26095164/184505353-2e6de3f2-40ac-480e-a9dd-c441691eedaf.png">
 
<img width="390" alt="Bildschirmfoto 2022-08-13 um 19 53 47" src="https://user-images.githubusercontent.com/26095164/184505354-531caaa2-d141-4c67-b6a1-11aa33942f71.png">
<img width="388" alt="Bildschirmfoto 2022-08-13 um 19 53 34" src="https://user-images.githubusercontent.com/26095164/184505355-fb04ce71-653c-4148-8f7d-8a62c9985e2e.png">
<img width="379" alt="Bildschirmfoto 2022-08-13 um 19 53 19" src="https://user-images.githubusercontent.com/26095164/184505356-7f58d8c9-f38e-42fd-9195-479414e7a52f.png">
<img width="434" alt="Bildschirmfoto 2022-08-13 um 19 53 09" src="https://user-images.githubusercontent.com/26095164/184505357-21ceb3f3-9f52-49af-a965-37714ddb83d1.png">
<img width="400" alt="Bildschirmfoto 2022-08-13 um 19 52 57" src="https://user-images.githubusercontent.com/26095164/184505358-ac60d2f7-4144-419d-a1b8-ff628db7f155.png">
<img width="753" alt="Bildschirmfoto 2022-08-13 um 19 52 47" src="https://user-images.githubusercontent.com/26095164/184505359-c4ff6382-d2e9-4cbc-af89-4f2420ac54cc.png">
<img width="763" alt="Bildschirmfoto 2022-08-13 um 19 52 37" src="https://user-images.githubusercontent.com/26095164/184505361-d67b56cb-db9b-490d-9c3d-2c4253723f52.png">


<img width="793" alt="Bildschirmfoto 2022-08-13 um 19 52 23" src="https://user-images.githubusercontent.com/26095164/184505352-c57d2969-d945-4738-86d7-fe7290ac4eba.png">

**Reply by ghost** [reply]
<img width="300" alt="Bildschirmfoto 2022-08-13 um 19 59 05" src="https://user-images.githubusercontent.com/412645/184505531-ac2a6096-3ebe-4544-984b-c36dfca905db.png">

Uncheck "Lock aspect ratio" and setup the height to the orange lines. All ROIs should have the same y

**Reply by danieldownload** [reply]
I'll try

**Reply by ghost** [reply]
The difference of the positions of the digits can you ignore. The model will recognize 0.1 or 9.9, but the post process calculate the zero.

**Reply by danieldownload** [reply]
like that?

<img width="700" alt="Bildschirmfoto 2022-08-13 um 20 07 59" src="https://user-images.githubusercontent.com/26095164/184505752-af6fc7fe-3b7b-416e-ac4d-c0fc4ec20413.png">

**Reply by ghost** [reply]
You could set the "ExtendedResolution" in configuration to true. So you have a digit more in the value, especially if your last digit moves very slowly.

**Reply by ghost** [reply]
NO! Like that:
![image](https://user-images.githubusercontent.com/412645/184505960-36273733-3b9e-474e-b916-2ddab69486ce.png)

**Reply by ghost** [reply]
The outer lines fits the top and bottom of the digit window.

**Reply by danieldownload** [reply]
ok

**Reply by danieldownload** [reply]
ok, realigned everything, now it looks like this:

Gonna change the last digit config to Extendet now.

!!!The overexposed image (2) still appears after reboot, only changes when I re-Reference the Image!!! So there is something...

Why seem the alignments on Overview page shifted to left?

<img width="1136" alt="Bildschirmfoto 2022-08-13 um 20 36 23" src="https://user-images.githubusercontent.com/26095164/184506608-549266be-701d-40b2-bf0f-3cbcfffaedc4.png">
<img width="1130" alt="Bildschirmfoto 2022-08-13 um 20 34 22" src="https://user-images.githubusercontent.com/26095164/184506614-8aeb9232-019a-4d18-bbcf-93622dbba549.png">
<img width="376" alt="Bildschirmfoto 2022-08-13 um 20 34 02" src="https://user-images.githubusercontent.com/26095164/184506615-178e4d0d-2815-423c-8a22-b39b3ac64d1a.png">
<img width="367" alt="Bildschirmfoto 2022-08-13 um 20 33 49" src="https://user-images.githubusercontent.com/26095164/184506618-b2b4e42f-ad0f-4612-b005-f8b43ba6ee43.png">
<img width="370" alt="Bildschirmfoto 2022-08-13 um 20 33 38" src="https://user-images.githubusercontent.com/26095164/184506619-8a738c9a-3f68-4250-9acb-11d3d26a1820.png">
<img width="370" alt="Bildschirmfoto 2022-08-13 um 20 33 28" src="https://user-images.githubusercontent.com/26095164/184506620-938b5f24-ba85-423c-a5ae-4fd72f8a9b82.png">
<img width="371" alt="Bildschirmfoto 2022-08-13 um 20 33 15" src="https://user-images.githubusercontent.com/26095164/184506621-6ee8b44a-8c50-43cf-a910-96f596a7fa61.png">
<img width="368" alt="Bildschirmfoto 2022-08-13 um 20 33 05" src="https://user-images.githubusercontent.com/26095164/184506622-8cafcbcf-c744-4960-badb-1b45adb97f66.png">
<img width="372" alt="Bildschirmfoto 2022-08-13 um 20 32 53" src="https://user-images.githubusercontent.com/26095164/184506623-8e85a4d5-c799-4194-b207-17c604e8f76e.png">
<img width="372" alt="Bildschirmfoto 2022-08-13 um 20 32 42" src="https://user-images.githubusercontent.com/26095164/184506624-3ff5d37d-51ea-4ca3-b1b8-7d53cee9a700.png">

**Reply by ghost** [reply]
The alignment depends on the images you configured. Could be depending on the higher brightness or you could set the alignment pictures again.
The digits could be a bit wider. Horizontally the digit should fit left/right the inner rectangle.

**Reply by ghost** [reply]
Normally you should not have an "N" any more. Wich model you use?

**Reply by danieldownload** [reply]
dig_class11_1411 right now

**Reply by ghost** [reply]
Better you use dig-class100 or dig-cont

**Reply by danieldownload** [reply]
ok lets try

**Reply by ghost** [reply]
DecimalShift should be -3 in your case

**Reply by danieldownload** [reply]
dig-cont shows:

<img width="1143" alt="Bildschirmfoto 2022-08-13 um 21 02 20" src="https://user-images.githubusercontent.com/26095164/184507338-27cb70c2-e5a7-4f74-9114-a04cfdecadac.png">

**Reply by ghost** [reply]
Looks better. But the red 6 is not ok. You could try the dig-class100 if its better and could play with alignment setup.

But I think the red digits are the digits after the decimal point. See DecimalShift above.

**Reply by danieldownload** [reply]
dig-class100 shows this, I think I will go with dig-class11 and remove the last digit, then it's exact on 10 liter.

<img width="1154" alt="Bildschirmfoto 2022-08-13 um 21 13 12" src="https://user-images.githubusercontent.com/26095164/184507643-a262e295-3092-4954-98e1-5b024d54c2f4.png">

**Reply by danieldownload** [reply]
DigitShift is already on -3

**Reply by ghost** [reply]
Instead of "C" I would try to use the qr-code for alignement.

**Reply by danieldownload** [reply]
ok, lets change it

**Reply by ghost** [reply]
I saw every time has the alignment the same shift. Move 2 or 3 dots in x less for every digit.

**Reply by danieldownload** [reply]
Moved from C to QR and every digit -3 pix. Result looks good now. I notice the -3 pix seem already too much now, is it because of the change of the alignment point?

<img width="1144" alt="Bildschirmfoto 2022-08-13 um 21 28 49" src="https://user-images.githubusercontent.com/26095164/184508123-149ee7b6-00c9-452f-9dc3-45798d5354b8.png">

**Reply by danieldownload** [reply]
Also m3 alignment point seems to have moved but I did not touch it...

**Reply by ghost** [reply]
I would try to set the m3 alignment point again. After that it could be the ROIs must moved again. Normally after every take of reference image the alignment and the ROIs must be configured again.

**Reply by danieldownload** [reply]
It moved back by itself :) strange but the values are correct. I will observe it the next two days and play then with the cam alignment, at the moment the cam itself has around 45 degree rotation.

<img width="1154" alt="Bildschirmfoto 2022-08-13 um 21 40 44" src="https://user-images.githubusercontent.com/26095164/184508438-b5020d36-c27a-4303-ae5c-c2e20a441446.png">

**Reply by danieldownload** [reply]
one more pic after switching to 4 at the last digit, very accurate.

<img width="1155" alt="Bildschirmfoto 2022-08-13 um 21 43 33" src="https://user-images.githubusercontent.com/26095164/184508543-4094a7cf-1d68-4eb3-a678-a77976329503.png">

**Reply by danieldownload** [reply]
Is it possible to hide the 000 at the beginning of the digital number?

**Reply by ghost** [reply]
Now set the previous value to the current ant next time the "value" have the number without trailing 000

**Reply by danieldownload** [reply]
The pic on Overview is somehow moving from shot to shot, or at least the alignments and data points. I set the Previous Value to 16.60xx but the zeros in front are still there.

<img width="1152" alt="Bildschirmfoto 2022-08-13 um 21 58 20" src="https://user-images.githubusercontent.com/26095164/184509020-b27cae89-b948-4cee-8a6e-158f434614d7.png">

**Reply by danieldownload** [reply]
just after last refresh on Overview, didn't touch the meter...

<img width="1141" alt="Bildschirmfoto 2022-08-13 um 22 02 15" src="https://user-images.githubusercontent.com/26095164/184509159-a967744e-214f-4164-8eb9-2e71ec33774f.png">

**Reply by danieldownload** [reply]
one more refresh and it's back to normal with correct values.

<img width="1129" alt="Bildschirmfoto 2022-08-13 um 22 06 22" src="https://user-images.githubusercontent.com/26095164/184509195-d81df922-0b6f-4087-9b39-c16d510adeaa.png">

**Reply by ghost** [reply]
Look at "Value:" its now without trailing zeros.

Could it be you set the alignment only for the qr-code? Configure both again and store it. If you look at the screenshot above the qr-code is not centered. Means the m3 and the qr-code can not be aligned together. 

If nothing helps. Begin with reference image (new reference and don't forget to store it). After both alignments and last look at the ROI-Configuration again. After all reboot and it should fit after a few minutes.

**Reply by ghost** [reply]
<img width="366" alt="image" src="https://user-images.githubusercontent.com/412645/184509215-bfc8f309-7b0f-4a25-bf2d-4ec46401d65e.png">
See at status. The screenshot is before alignment.

**Reply by danieldownload** [reply]
I rearranged the alignment points, the QR code box is a bit shifted to the right, compared as shown on Reference Image page, while m3 looks correct. Gonna turn the cam tomorrow to look straight on the number, it must have something to do with the turn and the angle in which the cam looks down on the numbers.

Looks very promising, thanks a million for your help 👍

<img width="734" alt="Bildschirmfoto 2022-08-13 um 22 18 08" src="https://user-images.githubusercontent.com/26095164/184509604-0c03bfa7-a7d7-498c-b87f-7963a7be777d.png">

<img width="749" alt="Bildschirmfoto 2022-08-13 um 22 23 25" src="https://user-images.githubusercontent.com/26095164/184509642-c7a3b5a3-7e9c-4c8b-bfa5-9a0a6e1a35b9.png">

**Reply by danieldownload** [reply]
ahhhh, just noticed I didn't take a new pic for referencing...

**Comment by rspring** [comment]
I came across this older discussion. Still I would like to share my solution with you. I have a water meter with bright red background on last three digits resulting in poor recognition performance. Now I am using an external RGB LED with color set in the configuration to RGB(0, 32, 32). This almost completely turns the bright red into dark black :) From this:
![image](https://user-images.githubusercontent.com/6276750/194061565-4cf3b3e2-7827-4f9d-b49b-732d01cdff9d.png)
Into this:
![image](https://user-images.githubusercontent.com/6276750/194058476-845f04b5-1f36-4d95-91b9-53a08fa6f641.png)

**Reply by caco3** [reply]
I have a RGB ambiente light at home, turning it to red also makes all red things to nearly look black, really impressive

**Comment by whitedavidp** [comment]
Thanks to those who spoke to changing the light color. At this point, I am pretty much sticking with the onboard LED. But the idea came to me that I could filter its light into the green and maybe get the same sort of effect.

I have my camera module in [one of these nice cases](https://www.etsy.com/listing/1895489791/esp32-cam-module-case-with-gopro-mount). I was able to cut a double layer of [something like this](https://www.amazon.com/JAM-PAPER-Plastic-Sleeves-Protectors/dp/B07HM5L7GF?crid=2HXO8VQV7VEYT&dib=eyJ2IjoiMSJ9.HYWAEHZGQKQmxXOD4p4oSdVsZNFj1wBMbNwI3DFVJdcfzWlrLPPA-AWjWUyb-_sAf7wpyYUbQQ3q2MQGb_kku3tI1DD9167y3IIa49fULjz08z_ebqdo4DLs2PJednGoeeSDrpbW1baMKUBqkfucp3Jj0CVQ7Mnpf82rq-N1zwzfLLOZRu5o1gwH7kGbu5OONVpRL9KpigQuHPKsfYH2DN4Op2uVA28onMka73fOXz4teeUpakiMvVDWGGDPKftJpJProbz9gSNjqGw-P_tSxvv8EZa_LlohQB0Dpwd74yY.rnS27YUhu4vl1Wu4HyloTCiy2NocnGgRWV7YxfTFGUE&dib_tag=se&keywords=multi-coloredplastic%2Bpaper%2Bsleeves&qid=1748105244&sprefix=multi-coloredplastic%2Bpaper%2Bsleeves%2Caps%2C189&sr=8-1&th=1) and place it between the camera board and the opening in the case for the LED. A little bit of a dance but not too bad. Just held in place by the pressure of the board against the case.

So far, this appears to make the photos being taken of the red colored last two digits on my meter more clear and get read more accurately.

Thanks
